src/builder.rs

```diff
@@ -0,0 +1,472 @@
+    value::{self, NoteValue, ValueCommitTrapdoor, ValueCommitment, ValueSum},
+    Address, EncryptedNote, Note,
+};
+
+const MIN_ACTIONS: usize = 2;
+
+#[derive(Debug)]
+pub enum Error {
+    MissingSignatures,
+    Proof(halo2::plonk::Error),
+    ValueSum(value::OverflowError),
+}
+
+impl From<halo2::plonk::Error> for Error {
+    fn from(e: halo2::plonk::Error) -> Self {
+        Error::Proof(e)
+    }
+}
+
+impl From<value::OverflowError> for Error {
+    fn from(e: value::OverflowError) -> Self {
+        Error::ValueSum(e)
+    }
+}
+
+/// Information about a specific note to be spent in an [`Action`].
+#[derive(Debug)]
+struct SpendInfo {
+    dummy_sk: Option<SpendingKey>,
+    fvk: FullViewingKey,
+    note: Note,
+    merkle_path: MerklePath,
+}
+
+impl SpendInfo {
+    /// Defined in [Zcash Protocol Spec § 4.8.3: Dummy Notes (Orchard)][orcharddummynotes].
+    ///
+    /// [orcharddummynotes]: https://zips.z.cash/protocol/nu5.pdf#orcharddummynotes
+    fn dummy(rng: &mut impl RngCore) -> Self {
+        let (sk, fvk, note) = Note::dummy(rng, None);
+        let merkle_path = MerklePath::dummy(rng);
+
+        SpendInfo {
+            dummy_sk: Some(sk),
+            fvk,
+            note,
+            merkle_path,
+        }
+    }
+}
+
+/// Information about a specific recipient to receive funds in an [`Action`].
+#[derive(Debug)]
+struct RecipientInfo {
+    ovk: Option<OutgoingViewingKey>,
+    recipient: Address,
+    value: NoteValue,
+    memo: Option<()>,
+}
+
+impl RecipientInfo {
+    /// Defined in [Zcash Protocol Spec § 4.8.3: Dummy Notes (Orchard)][orcharddummynotes].
+    ///
+    /// [orcharddummynotes]: https://zips.z.cash/protocol/nu5.pdf#orcharddummynotes
+    fn dummy(rng: &mut impl RngCore) -> Self {
+        let fvk: FullViewingKey = (&SpendingKey::random(rng)).into();
+        let recipient = fvk.default_address();
+
+        RecipientInfo {
+            ovk: None,
+            recipient,
+            value: NoteValue::default(),
+            memo: None,
+        }
+    }
+}
+
+/// Information about a specific [`Action`] we plan to build.
+#[derive(Debug)]
+struct ActionInfo {
+    spend: SpendInfo,
+    output: RecipientInfo,
+    rcv: ValueCommitTrapdoor,
+}
+
+impl ActionInfo {
+    fn new(spend: SpendInfo, output: RecipientInfo, rng: impl RngCore) -> Self {
+        ActionInfo {
+            spend,
+            output,
+            rcv: ValueCommitTrapdoor::random(rng),
+        }
+    }
+
+    /// Returns the value sum for this action.
+    fn value_sum(&self) -> Result<ValueSum, value::OverflowError> {
+        self.spend.note.value() - self.output.value
+    }
+
+    /// Builds the action.
+    ///
+    /// Defined in [Zcash Protocol Spec § 4.7.3: Sending Notes (Orchard)][orchardsend].
+    ///
+    /// [orchardsend]: https://zips.z.cash/protocol/nu5.pdf#orchardsend
+    fn build(
+        self,
+        mut rng: impl RngCore,
+    ) -> (
+        Action<(Option<SpendAuthorizingKey>, SpendValidatingKey)>,
```

Comment:
It's not obvious that this `SpendAuthorizingKey` will only be present if it's a dummy value. This at least needs documentation, but I feel like it might be better served by a bespoke enum type.

Reply:
Alternately: instead of a tuple, return a struct? I just want somewhere that the documentation for that first element can live.